### PR TITLE
pyload: init at stable

### PR DESCRIPTION
**Pull request by 2chilled**
###### Motivation for this change

Provide a nice download tool on nix
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -19260,11 +19260,16 @@ in modules // {
...
 
-    propagatedBuildInputs = with self; [ pkgs.curl pkgs.openssl ];
+    propagatedBuildInputs = with self; [ pkgs.curl pkgs.openssl.out ];
 
     # error: invalid command 'test'
     doCheck = false;
 
+    #TODO no idea why this is needed
+    postInstall = ''
+      ln -s ${pkgs.openssl.out}/lib/libcrypto* $out/lib/
```
Comment (Mic92) [inline pkgs/top-level/python-packages.nix]: Without this patch libopenssl is not found at runtime? Could you show me the output of `ldd`, in case this package contains a shared library.

File: pkgs/applications/networking/pyload/default.nix [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  patches =
+    # gets merged in next release version of pyload
+    let configParserPatch = fetchurl {
+      url = "https://patch-diff.githubusercontent.com/raw/pyload/pyload/pull/2625.patch";
+      sha256 = "0v8l840yxgzlbcxnm7cd4ipc52bbs798lnbyd6f9nrd85ih1nsfl";
+    };
+    in [configParserPatch];
+
+  pythonWithPkgs = (python.withPackages (ps: with ps; [
```
Comment (FRidh) [inline pkgs/applications/networking/pyload/default.nix]: this needs to be in a let expression instead of passing it as an argument to mkDerivation

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -19260,11 +19260,16 @@ in modules // {
       sha256 = "0hqsap82zklhi5fxhc69kxrwzb0g9566f7sdpz7f9gyxkmyam839";
     };
 
-    propagatedBuildInputs = with self; [ pkgs.curl pkgs.openssl ];
+    propagatedBuildInputs = with self; [ pkgs.curl pkgs.openssl.out ];
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: changes here need to be in a separate commit

File: pkgs/applications/networking/pyload/default.nix (hunk 2 above)
Comment (FRidh) [inline pkgs/applications/networking/pyload/default.nix]: also, while not a requirement, I try to stick with the name `pythonEnv` throughout Nixpkgs when creating Python environments with packages.

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/top-level/python-packages.nix (hunk 1 above)
Comment (2chilled) [inline pkgs/top-level/python-packages.nix]:
```
ldd result/lib/python2.7/site-packages/pycurl.so 
        linux-vdso.so.1 (0x00007ffe4ff9a000)
        libgcc_s.so.1 => /nix/store/23vndix3mzcyw84kag06h8fh464g32nq-glibc-2.24/lib/libgcc_s.so.1 (0x00007f492376d000)
        libcurl.so.4 => /nix/store/wl6zyv8102v5c0wafxyp665y1xqca6ay-curl-7.50.3/lib/libcurl.so.4 (0x00007f49234f9000)
        libpython2.7.so.1.0 => /nix/store/18ns8aqaglqh7k0fn2cqih8kc29qsryg-python-2.7.12/lib/libpython2.7.so.1.0 (0x00007f492311b000)
        libpthread.so.0 => /nix/store/23vndix3mzcyw84kag06h8fh464g32nq-glibc-2.24/lib/libpthread.so.0 (0x00007f4922efe000)
        libc.so.6 => /nix/store/23vndix3mzcyw84kag06h8fh464g32nq-glibc-2.24/lib/libc.so.6 (0x00007f4922b5e000)
        libcrypto.so.1.0.0 => not found
        libnghttp2.so.14 => /nix/store/49pvdx6m91inzr0afrd89n9p6i1r2xz3-nghttp2-1.14.1-lib/lib/libnghttp2.so.14 (0x00007f4922939000)
        libssh2.so.1 => /nix/store/507j9c95hhyg2pm1arpjimyfcf6q022n-libssh2-1.7.0/lib/libssh2.so.1 (0x00007f492270b000)
        libssl.so.1.0.0 => /nix/store/4vq58j5cdpwpxmw2272kp3g9l9g8h71v-openssl-1.0.2j/lib/libssl.so.1.0.0 (0x00007f492249d000)
        libcrypto.so.1.0.0 => /nix/store/4vq58j5cdpwpxmw2272kp3g9l9g8h71v-openssl-1.0.2j/lib/libcrypto.so.1.0.0 (0x00007f492205f000)
        libz.so.1 => /nix/store/caz03y58cyrqn4i5ckdw4i21g9sdjb18-zlib-1.2.8/lib/libz.so.1 (0x00007f4921e49000)
        libdl.so.2 => /nix/store/23vndix3mzcyw84kag06h8fh464g32nq-glibc-2.24/lib/libdl.so.2 (0x00007f4921c45000)
        libutil.so.1 => /nix/store/23vndix3mzcyw84kag06h8fh464g32nq-glibc-2.24/lib/libutil.so.1 (0x00007f4921a42000)
        libm.so.6 => /nix/store/23vndix3mzcyw84kag06h8fh464g32nq-glibc-2.24/lib/libm.so.6 (0x00007f492173d000)
        /nix/store/jm1n87rp8vr90j9ahcrfzr57nc2r8vgf-glibc-2.24/lib64/ld-linux-x86-64.so.2 (0x00005595a1f4b000)
```

File: pkgs/applications/networking/pyload/default.nix [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  ]));
+in
+stdenv.mkDerivation rec {
+  version = "stable";
+  name = "pyLoad-" + version;
+
+  src = fetchFromGitHub {
+    owner = "pyload";
+    repo = "pyload";
+    rev = "stable";
```
Comment (FRidh) [inline pkgs/applications/networking/pyload/default.nix]: This has to be a fixed revision, for example a tag / release. In the case of a tag / release, you can also use the tarball.

File: pkgs/applications/networking/pyload/default.nix [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  pythonEnv = (python.withPackages (ps: with ps; [
+    pycurl jinja2 beaker thrift simplejson
+    #optional packages
+    #TODO add Send2Trash
+    #TODO package the other optional ones, see homepage
+    pycrypto feedparser pyqt4
+  ]));
+in
+stdenv.mkDerivation rec {
+  version = "stable";
```
Comment (FRidh) [inline pkgs/applications/networking/pyload/default.nix]: According to the Nixpkgs guidelines `stable` is not a version. I suggest you use the latest release.

File: pkgs/applications/networking/pyload/default.nix [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  name = "pyLoad-" + version;
+
+  src = fetchFromGitHub {
+    owner = "pyload";
+    repo = "pyload";
+    rev = "stable";
+    sha256 = "13as5z38v4iaq9fziiq5cim3bmllwcf4dhpigh7mvrih93pn90hn";
+  };
+
+  installPhase =
```
Comment (FRidh) [inline pkgs/applications/networking/pyload/default.nix]: looking a bit further at the package, it seems it is a Python package that uses `paver`. We have `paver` in `python-packages.nix`. And it's possible to create a `setup.py` with `paver generate_setup`. Doing that should result in a `setup.py` allowing the package to be installed with `buildPythonApplication`.

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by FRidh** [comment]
I think https://github.com/FRidh/nixpkgs/tree/pyload is the correct approach (except the change fetchpatch -> fetchurl ). Haven't tested yet though.

**Comment by 2chilled** [comment]
@FRidh This looks promising. Although there's a missing piece as I get 
```
Traceback (most recent call last):
  File "/nix/store/g3piccnmjlfzqcphbnkzpr9yclgz7i2j-pyLoad-0.4.9/bin/.pyLoadCore-wrapped", line 8, in <module>
    from pyLoadCore import main
ImportError: No module named pyLoadCore
```
when running the bin/pyLoadCore. I'll take a look at that when I've more time.

**Comment by 2chilled** [comment]
@FRidh The tarball 0.4.9 which you are using in your proposed change is broken. According to this issue https://github.com/pyload/pyload/issues/624 it is recommended to use https://github.com/pyload/pyload/archive/stable.zip

**Comment by FRidh** [comment]
> @FRidh The tarball 0.4.9 which you are using in your proposed change is broken. According to this issue pyload/pyload#624 it is recommended to use https://github.com/pyload/pyload/archive/stable.zip

You cannot use a tarball that refers to the HEAD of a branch, since the contents changes whenever they commit something, and thus also the hash of the archive. You have to pick something that doesn't vary, like a specific commit, or a release/tag.

**Comment by 2chilled** [comment]
Then I have to use a specific commit since the youngest tag points to the broken version. Also, the pavement.py of pyload is incomplete which is why we have the mentioned import error above.

**Comment by 2chilled** [comment]
@FRidh I created a further pyload PR to be able to support your strategy using paver for being able to use buildPythonApplication. Both pyload PR diffs are included in the nix derivation.

**Comment by FRidh** [comment]
Great. And is it working now?

**Comment by 2chilled** [comment]
Yeah, finally :)
